# **ANN on Breast Cancer Dataset**

***Given breast cancer dataset and ANN model python notebook is given***

***Change different parameters to improve the model accuracy ***

In [1]:
#importing pandas
import pandas as pd

In [2]:
from google.colab import files
uploaded=files.upload()

Saving breastcancer.csv to breastcancer.csv


In [3]:
#reading csv file
data = pd.read_csv("breastcancer.csv")

In [6]:
data.head(5)

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  texture_worst  perimeter_worst  area_worst  smoothness_worst  \
0  ...          17.33           184.60      2019.0            0.1622   
1  ...          23.41           158.80      1956.0            0.1238   
2  ...          25.53           152.50      1709.0            0.1444   
3  ...          26.50            98.87       567.7            0.2098   
4  ...          16.67           152.20      1575.0            0.1374   

   compactness_worst  concavity_worst  concave points_worst  symmetry_worst  \
0             0.6656           0.7119                0.2654          0.4601   
1             0.1866           0.2416                0.1860          0.2750   
2             0.4245           0.4504                0.2430          0.3613   
3             0.8663           0.6869                0.2575          0.6638   
4             0.2050           0.4000                0.1625          0.2364   

   fractal_dimension_worst  Unnamed: 32  
0                  0.11890          NaN  
1                  0.08902          NaN  
2                  0.08758          NaN  
3                  0.17300          NaN  
4                  0.07678          NaN  

[5 rows x 33 columns]

In [7]:
#finding columns
data.columns

Index(['id', 'diagnosis', 'radius_mean', 'texture_mean', 'perimeter_mean',
       'area_mean', 'smoothness_mean', 'compactness_mean', 'concavity_mean',
       'concave points_mean', 'symmetry_mean', 'fractal_dimension_mean',
       'radius_se', 'texture_se', 'perimeter_se', 'area_se', 'smoothness_se',
       'compactness_se', 'concavity_se', 'concave points_se', 'symmetry_se',
       'fractal_dimension_se', 'radius_worst', 'texture_worst',
       'perimeter_worst', 'area_worst', 'smoothness_worst',
       'compactness_worst', 'concavity_worst', 'concave points_worst',
       'symmetry_worst', 'fractal_dimension_worst', 'Unnamed: 32'],
      dtype='object')

**Preprocessing**





In [8]:
#finding null values
data.isna().sum()

id                           0
diagnosis                    0
radius_mean                  0
texture_mean                 0
perimeter_mean               0
area_mean                    0
smoothness_mean              0
compactness_mean             0
concavity_mean               0
concave points_mean          0
symmetry_mean                0
fractal_dimension_mean       0
radius_se                    0
texture_se                   0
perimeter_se                 0
area_se                      0
smoothness_se                0
compactness_se               0
concavity_se                 0
concave points_se            0
symmetry_se                  0
fractal_dimension_se         0
radius_worst                 0
texture_worst                0
perimeter_worst              0
area_worst                   0
smoothness_worst             0
compactness_worst            0
concavity_worst              0
concave points_worst         0
symmetry_worst               0
fractal_dimension_worst      0
Unnamed:

In [9]:
#finding datatypes of the columns
data.dtypes

id                           int64
diagnosis                   object
radius_mean                float64
texture_mean               float64
perimeter_mean             float64
area_mean                  float64
smoothness_mean            float64
compactness_mean           float64
concavity_mean             float64
concave points_mean        float64
symmetry_mean              float64
fractal_dimension_mean     float64
radius_se                  float64
texture_se                 float64
perimeter_se               float64
area_se                    float64
smoothness_se              float64
compactness_se             float64
concavity_se               float64
concave points_se          float64
symmetry_se                float64
fractal_dimension_se       float64
radius_worst               float64
texture_worst              float64
perimeter_worst            float64
area_worst                 float64
smoothness_worst           float64
compactness_worst          float64
concavity_worst     

Only the column "diagnosis" is object, rest all are number types.

The "column "Unnamed" has null values, but this is not valid or required for the model. So, this can be dropped while defining the features as this feature is not required
The Target will be "diagnosis"

In [10]:
#defining features
x = data[[ 'radius_mean', 'texture_mean', 'perimeter_mean',
       'area_mean', 'smoothness_mean', 'compactness_mean', 'concavity_mean',
       'concave points_mean', 'symmetry_mean', 'fractal_dimension_mean',
       'radius_se', 'texture_se', 'perimeter_se', 'area_se', 'smoothness_se',
       'compactness_se', 'concavity_se', 'concave points_se', 'symmetry_se',
       'fractal_dimension_se', 'radius_worst', 'texture_worst',
       'perimeter_worst', 'area_worst', 'smoothness_worst',
       'compactness_worst', 'concavity_worst', 'concave points_worst',
       'symmetry_worst', 'fractal_dimension_worst']]

In [11]:
#defining target
y = data['diagnosis']

In [12]:
#check distinct values
y.value_counts()

B    357
M    212
Name: diagnosis, dtype: int64

diagnosis field, make it as boolean, M as 1 and B as 0

In [13]:
y  = y.map({"B":0,"M":1})

**Scaling of features** :

In [15]:
x.describe()

radius_mean  texture_mean  perimeter_mean    area_mean  \
count   569.000000    569.000000      569.000000   569.000000   
mean     14.127292     19.289649       91.969033   654.889104   
std       3.524049      4.301036       24.298981   351.914129   
min       6.981000      9.710000       43.790000   143.500000   
25%      11.700000     16.170000       75.170000   420.300000   
50%      13.370000     18.840000       86.240000   551.100000   
75%      15.780000     21.800000      104.100000   782.700000   
max      28.110000     39.280000      188.500000  2501.000000   

       smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
count       569.000000        569.000000      569.000000           569.000000   
mean          0.096360          0.104341        0.088799             0.048919   
std           0.014064          0.052813        0.079720             0.038803   
min           0.052630          0.019380        0.000000             0.000000   
25%           0.086370          0.064920        0.029560             0.020310   
50%           0.095870          0.092630        0.061540             0.033500   
75%           0.105300          0.130400        0.130700             0.074000   
max           0.163400          0.345400        0.426800             0.201200   

       symmetry_mean  fractal_dimension_mean  ...  radius_worst  \
count     569.000000              569.000000  ...    569.000000   
mean        0.181162                0.062798  ...     16.269190   
std         0.027414                0.007060  ...      4.833242   
min         0.106000                0.049960  ...      7.930000   
25%         0.161900                0.057700  ...     13.010000   
50%         0.179200                0.061540  ...     14.970000   
75%         0.195700                0.066120  ...     18.790000   
max         0.304000                0.097440  ...     36.040000   

       texture_worst  perimeter_worst   area_worst  smoothness_worst  \
count     569.000000       569.000000   569.000000        569.000000   
mean       25.677223       107.261213   880.583128          0.132369   
std         6.146258        33.602542   569.356993          0.022832   
min        12.020000        50.410000   185.200000          0.071170   
25%        21.080000        84.110000   515.300000          0.116600   
50%        25.410000        97.660000   686.500000          0.131300   
75%        29.720000       125.400000  1084.000000          0.146000   
max        49.540000       251.200000  4254.000000          0.222600   

       compactness_worst  concavity_worst  concave points_worst  \
count         569.000000       569.000000            569.000000   
mean            0.254265         0.272188              0.114606   
std             0.157336         0.208624              0.065732   
min             0.027290         0.000000              0.000000   
25%             0.147200         0.114500              0.064930   
50%             0.211900         0.226700              0.099930   
75%             0.339100         0.382900              0.161400   
max             1.058000         1.252000              0.291000   

       symmetry_worst  fractal_dimension_worst  
count      569.000000               569.000000  
mean         0.290076                 0.083946  
std          0.061867                 0.018061  
min          0.156500                 0.055040  
25%          0.250400                 0.071460  
50%          0.282200                 0.080040  
75%          0.317900                 0.092080  
max          0.663800                 0.207500  

[8 rows x 30 columns]

In [16]:
#Min Max scaler
from sklearn import preprocessing
min_max=preprocessing.MinMaxScaler(feature_range=(0,1))
x=min_max.fit_transform(x)
x=pd.DataFrame(x)
x.describe()

0           1           2           3           4           5   \
count  569.000000  569.000000  569.000000  569.000000  569.000000  569.000000   
mean     0.338222    0.323965    0.332935    0.216920    0.394785    0.260601   
std      0.166787    0.145453    0.167915    0.149274    0.126967    0.161992   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.223342    0.218465    0.216847    0.117413    0.304595    0.139685   
50%      0.302381    0.308759    0.293345    0.172895    0.390358    0.224679   
75%      0.416442    0.408860    0.416765    0.271135    0.475490    0.340531   
max      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   

               6           7           8           9   ...          20  \
count  569.000000  569.000000  569.000000  569.000000  ...  569.000000   
mean     0.208058    0.243137    0.379605    0.270379  ...    0.296663   
std      0.186785    0.192857    0.138456    0.148702  ...    0.171940   
min      0.000000    0.000000    0.000000    0.000000  ...    0.000000   
25%      0.069260    0.100944    0.282323    0.163016  ...    0.180719   
50%      0.144189    0.166501    0.369697    0.243892  ...    0.250445   
75%      0.306232    0.367793    0.453030    0.340354  ...    0.386339   
max      1.000000    1.000000    1.000000    1.000000  ...    1.000000   

               21          22          23          24          25          26  \
count  569.000000  569.000000  569.000000  569.000000  569.000000  569.000000   
mean     0.363998    0.283138    0.170906    0.404138    0.220212    0.217403   
std      0.163813    0.167352    0.139932    0.150779    0.152649    0.166633   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.241471    0.167837    0.081130    0.300007    0.116337    0.091454   
50%      0.356876    0.235320    0.123206    0.397081    0.179110    0.181070   
75%      0.471748    0.373475    0.220901    0.494156    0.302520    0.305831   
max      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   

               27          28          29  
count  569.000000  569.000000  569.000000  
mean     0.393836    0.263307    0.189596  
std      0.225884    0.121954    0.118466  
min      0.000000    0.000000    0.000000  
25%      0.223127    0.185098    0.107700  
50%      0.343402    0.247782    0.163977  
75%      0.554639    0.318155    0.242949  
max      1.000000    1.000000    1.000000  

[8 rows x 30 columns]

**Splitting of Test and Train values:**

In [17]:
from sklearn.model_selection import train_test_split

In [18]:
x_train, x_test, y_train, y_test = train_test_split(x,y,
                                                     test_size=0.2,
                                                     random_state = 134)

In [19]:
x_train.shape

(455, 30)

**ANN**

In [20]:
#import tensorflow
import tensorflow as tf

In [21]:
#check the version of tensorflow
tf.__version__

'2.11.0'

In [23]:
#import keras models
from tensorflow.keras.models import Sequential

In [24]:
#import keras layers
from tensorflow.keras.layers import Dense

# **Model 1**

add layers (hidden layer 1,2,3.. and output layer)

**First Step : Create Architecture**

Create architecture

input layer - 30 features of x

hidden layer 1-8

output layer = 1, as it is binary classification 

create the instance of Sequential model

In [25]:
model1 = Sequential()
model1.add(Dense(32, input_shape = (30,), activation = "relu"))
model1.add(Dense(32, activation = "relu"))
model1.add(Dense(1, activation = "sigmoid"))
model1.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 32)                992       
                                                                 
 dense_1 (Dense)             (None, 32)                1056      
                                                                 
 dense_2 (Dense)             (None, 1)                 33        
                                                                 
Total params: 2,081
Trainable params: 2,081
Non-trainable params: 0
_________________________________________________________________


**Second Step : Compile**

In [26]:
model1.compile(optimizer = "sgd", loss = "binary_crossentropy",
               metrics = ["accuracy"])

**Third Step: Fitting**

In [27]:
model1.fit(x_train, y_train, epochs = 10)

Epoch 1/10
15/15 [==============================] - 4s 4ms/step - loss: 0.6857 - accuracy: 0.6308
Epoch 2/10
15/15 [==============================] - 0s 5ms/step - loss: 0.6798 - accuracy: 0.6308
Epoch 3/10
15/15 [==============================] - 0s 5ms/step - loss: 0.6742 - accuracy: 0.6308
Epoch 4/10
15/15 [==============================] - 0s 6ms/step - loss: 0.6684 - accuracy: 0.6308
Epoch 5/10
15/15 [==============================] - 0s 4ms/step - loss: 0.6626 - accuracy: 0.6308
Epoch 6/10
15/15 [==============================] - 0s 4ms/step - loss: 0.6569 - accuracy: 0.6308
Epoch 7/10
15/15 [==============================] - 0s 4ms/step - loss: 0.6511 - accuracy: 0.6308
Epoch 8/10
15/15 [==============================] - 0s 5ms/step - loss: 0.6460 - accuracy: 0.6308
Epoch 9/10
15/15 [==============================] - 0s 4ms/step - loss: 0.6409 - accuracy: 0.6308
Epoch 10/10
15/15 [==============================] - 0s 4ms/step - loss: 0.6357 - accuracy: 0.6308


**Fourth Step : Predictions**

In [28]:
model1.evaluate(x_test,y_test) 

4/4 [==============================] - 0s 4ms/step - loss: 0.6335 - accuracy: 0.6140


[0.6334975957870483, 0.6140350699424744]

So, using Model1, we got the accuracy as 61%

# **Model 2**

**First Step: Create Architecture**

In [30]:
#input=30, hidden1=16, hidden2=16,output =1
model2=Sequential()
model2.add(Dense(16, input_shape=(30,),activation ='relu'))
model2.add(Dense(16,activation ='relu'))
model2.add(Dense(1,activation ='relu'))
model2.summary()

Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_6 (Dense)             (None, 16)                496       
                                                                 
 dense_7 (Dense)             (None, 16)                272       
                                                                 
 dense_8 (Dense)             (None, 1)                 17        
                                                                 
Total params: 785
Trainable params: 785
Non-trainable params: 0
_________________________________________________________________


**Second step : Compile**

In [31]:
model2.compile(optimizer = "sgd", loss = "binary_crossentropy",
               metrics = ["accuracy"])

**Third step : Fitting**

In [32]:
model2.fit(x_train, y_train, epochs = 10)

Epoch 1/10
15/15 [==============================] - 1s 3ms/step - loss: 5.6954 - accuracy: 0.6308
Epoch 2/10
15/15 [==============================] - 0s 4ms/step - loss: 5.6954 - accuracy: 0.6308
Epoch 3/10
15/15 [==============================] - 0s 3ms/step - loss: 5.6954 - accuracy: 0.6308
Epoch 4/10
15/15 [==============================] - 0s 3ms/step - loss: 5.6954 - accuracy: 0.6308
Epoch 5/10
15/15 [==============================] - 0s 3ms/step - loss: 5.6954 - accuracy: 0.6308
Epoch 6/10
15/15 [==============================] - 0s 3ms/step - loss: 5.6954 - accuracy: 0.6308
Epoch 7/10
15/15 [==============================] - 0s 3ms/step - loss: 5.6954 - accuracy: 0.6308
Epoch 8/10
15/15 [==============================] - 0s 3ms/step - loss: 5.6954 - accuracy: 0.6308
Epoch 9/10
15/15 [==============================] - 0s 3ms/step - loss: 5.6954 - accuracy: 0.6308
Epoch 10/10
15/15 [==============================] - 0s 4ms/step - loss: 5.6954 - accuracy: 0.6308


**Fourth Step : Predictions**

In [33]:
model2.evaluate(x_test,y_test) 

4/4 [==============================] - 0s 5ms/step - loss: 5.9535 - accuracy: 0.6140


[5.953489303588867, 0.6140350699424744]

Second model also gave an accuracy of 61.4%

# **Model 3**

**First Step: Create Architecture**

In [43]:
from keras.layers import Dropout
classifier=Sequential()
classifier.add(Dense(16, input_shape=(30,),activation ='relu',kernel_initializer='uniform'))
# Adding dropout to prevent overfitting
#classifier.add(Dropout(p=0.1))
classifier.add(Dropout(0.2, input_shape=(30,)))


# Adding the second hidden layer
classifier.add(Dense(16, kernel_initializer='uniform', activation='relu'))
# Adding dropout to prevent overfitting
#classifier.add(Dropout(p=0.1))
classifier.add(Dropout(0.2, input_shape=(30,)))

# Adding the output layer (output_dim is 1 as we want only 1 output from the final layer.)
classifier.add(Dense(1, kernel_initializer='uniform', activation='sigmoid'))

In [44]:
classifier.summary()

Model: "sequential_12"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_15 (Dense)            (None, 16)                496       
                                                                 
 dropout_1 (Dropout)         (None, 16)                0         
                                                                 
 dense_16 (Dense)            (None, 16)                272       
                                                                 
 dropout_2 (Dropout)         (None, 16)                0         
                                                                 
 dense_17 (Dense)            (None, 1)                 17        
                                                                 
Total params: 785
Trainable params: 785
Non-trainable params: 0
_________________________________________________________________


**Second step : Compile**

In [45]:
classifier.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

**Third step : Fitting**

In [49]:
classifier.fit(x_train, y_train, batch_size=100, epochs=150)

Epoch 1/150
5/5 [==============================] - 1s 5ms/step - loss: 0.6928 - accuracy: 0.6066
Epoch 2/150
5/5 [==============================] - 0s 4ms/step - loss: 0.6919 - accuracy: 0.6308
Epoch 3/150
5/5 [==============================] - 0s 4ms/step - loss: 0.6909 - accuracy: 0.6308
Epoch 4/150
5/5 [==============================] - 0s 4ms/step - loss: 0.6896 - accuracy: 0.6330
Epoch 5/150
5/5 [==============================] - 0s 4ms/step - loss: 0.6881 - accuracy: 0.6374
Epoch 6/150
5/5 [==============================] - 0s 5ms/step - loss: 0.6862 - accuracy: 0.6462
Epoch 7/150
5/5 [==============================] - 0s 4ms/step - loss: 0.6841 - accuracy: 0.6681
Epoch 8/150
5/5 [==============================] - 0s 6ms/step - loss: 0.6805 - accuracy: 0.7429
Epoch 9/150
5/5 [==============================] - 0s 5ms/step - loss: 0.6775 - accuracy: 0.7692
Epoch 10/150
5/5 [==============================] - 0s 6ms/step - loss: 0.6719 - accuracy: 0.8440
Epoch 11/150
5/5 [===========

**Fourth Step : Predictions**

In [50]:
classifier.evaluate(x_test,y_test) 

4/4 [==============================] - 0s 4ms/step - loss: 0.0864 - accuracy: 0.9825


[0.08637237548828125, 0.9824561476707458]

The last model shows an accuracy of 98.25% when dropout was introduced to prevent overfirtting